src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,103 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] values;
+    private int size;
+
+    public ArrayList() {
+        values = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        isNeadToResize();
+        values[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        isNeadToResize();
+        if (index != size) {
+            checkIndexWithSameSize(index);
+            System.arraycopy(values, index, values, index + 1, size - index);
+        }
+        values[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexWithSameSize(index);
+        return values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexWithSameSize(index);
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexWithSameSize(index);
+        final T removedElement = values[index];
+        System.arraycopy(values, index + 1,
+                values, index,
+                values.length - index - 1);
+        values[--size] = null;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (isEquals(element, values[i])) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Element not found in ArrayList");
+    }
+
+    private void grow() {
+        int newCapacity = (int) (values.length * 1.5);
+        T[] newValues = (T[]) new Object[newCapacity];
+        System.arraycopy(values, 0, newValues, 0, size);
+        values = newValues;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void isNeadToResize() {
+        if (size == values.length || (size - 1) == values.length) {
+            grow();
+        }
+    }
+
+    private void checkIndexWithSameSize(int index) {
```

Comment:
I'd recommend renaming it to simple 'checkIndex' or 'validateIndex'

Reply:
thanks for all recommendation